### PR TITLE
refactor(android): temp file handling

### PR DESCRIPTION
**Pull request by jquick-axway**
**JIRA:**
- https://jira.appcelerator.org/browse/TIMOB-28058
- https://jira.appcelerator.org/browse/TIMOB-28176
- https://jira.appcelerator.org/browse/TIMOB-28177

**Summary:**
- Moved `Ti.Filesystem.tempDirectory` location:
  * From: `file:///.../<PackageName>/cache/_tmp`
  * To: `file:///.../<PackageName>/cache/.titanium/tmp`
- Moved `Ti.Filesystem.createTempFile()` location:
  * From: `Ti.Filesystem.applicationCacheDirectory`
  * To: `Ti.Filesystem.tempDirectory` _(Matches iOS' behavior.)_
- Moved `Ti.Filesystem.createTempDirectory()` location:
  * From a directory Titanium did not provide a constant for.
  * To: `Ti.Filesystem.tempDirectory` _(Matches iOS' behavior.)_
- Modified HTTP response cache file handling:
  * Moved from external storage to internal storage for privacy. _(Recommended by Google.)_
  * New location: `file:///.../<PackageName>/cache/.titanium/http-response-cache`
  * Supports caching if app does not have `WRITE_EXTERNAL_STORAGE` permission. _(This is new.)_
  * Cache no longer auto-deleted when re-launching app. _(Is limited to 25 MB.)_
  * Removed external storage mounting/unmounting monitoring. _(No longer needed.)_
- Simplified temp file deletion via hidden "trash" folder.
  * Old approach used to auto-delete files under temp folder 5 seconds after JS runtime starts, which means app can briefly find previous temp files at startup.
  * New approach "moves" temp folder under a hidden "trash" folder (this is very fast) and then the "trash" folder is deleted via a thread. A new temp folder will be created when requested in JS code. Temp files are trashed every time JS runtime terminates and when app starts up natively before JS runtime starts.
- Removed `TiTempFileHelper` Java class.

---
**Test:**
1. Build and run the below on Android.
2. On app startup, verify temp files found is: `<None>`
3. Tap on the "Create Temp Files" button.
4. Verify you see 3 files listed:
`file:///.../cache/.titanium/tmp/RootTempFile.txt`
`file:///.../cache/.titanium/tmp/tifile<Numbers>.tmp`
`file:///.../cache/.titanium/tmp/tidir<Numbers>/SubdirectoryFile.txt`
5. Tap on the "Create Temp Files" button again.
6. Verify you see 5 files listed: _(A new tifile and tidir will be generated.)_
`file:///.../cache/.titanium/tmp/RootTempFile.txt`
`file:///.../cache/.titanium/tmp/tifile<Numbers>.tmp`
`file:///.../cache/.titanium/tmp/tifile<Numbers>.tmp`
`file:///.../cache/.titanium/tmp/tidir<Numbers>/SubdirectoryFile.txt`
`file:///.../cache/.titanium/tmp/tidir<Numbers>/SubdirectoryFile.txt`
7. Back out of the app and quickly relaunch.
8. Verify temp files found is: `<None>`

```javascript
let label = null;
function readTempFiles() {
	let text = "";
	function loadFrom(file) {
		if (file.isDirectory()) {
			for (const nextFileName of file.getDirectoryListing()) {
				loadFrom(Ti.Filesystem.getFile(file.nativePath, nextFileName));
			}
		} else {
			text += "- " + file.nativePath + "\n";
		}
	}
	loadFrom(Ti.Filesystem.getFile(Ti.Filesystem.tempDirectory));
	label.text = "Temp Files Found:\n" + (text ? text : "<None>");
	console.log("@@@ " + label.text);
}

const window = Ti.UI.createWindow();
label = Ti.UI.createLabel({
	top: "5dp",
	left: "5dp",
	bottom: "5dp",
	right: "5dp",
});
window.add(label);
const button = Ti.UI.createButton({
	title: "Create Temp Files",
	bottom: "20dp",
});
button.addEventListener("click", function() {
	let tempFile = Ti.Filesystem.getFile(Ti.Filesystem.tempDirectory, "RootTempFile.txt");
	tempFile.write("Manually created file.")
	tempFile = Ti.Filesystem.createTempFile();
	tempFile.write("Generated file.");
	tempFile = Ti.Filesystem.getFile(Ti.Filesystem.createTempDirectory().nativePath, "SubdirectoryFile.txt");
	tempFile.write("Subdirectory file.");
	readTempFiles();
});
window.add(button);
window.addEventListener("open", function() {
	readTempFiles();
});
window.open();
```

### CODE REVIEW

File: android/titanium/src/java/org/appcelerator/titanium/TiApplication.java [hunk 1]
```diff
@@ -468,23 +467,158 @@ public void postOnCreate()
...
+		// If given a directory, then recursively delete the entire tree.
+		boolean wasDeleted = true;
+		if (file.isDirectory()) {
+			for (File nextFile : file.listFiles()) {
+				wasDeleted = deleteTree(nextFile) && wasDeleted;
+			}
+		}
+
+		// Delete the given directory/file.
+		return (file.delete() && wasDeleted);
```
Comment (sgtcoolguy) [inline android/titanium/src/java/org/appcelerator/titanium/TiApplication.java:621]: shouldn't we avoid trying to delete a folder if we know that a child already failed? (i.e. swap the order here)

**Review by sgtcoolguy — APPROVED** [review]
This is nice, looks like it simplifies things much more and feels like much less of a hassle to not need external storage/mounting to deal with here.

File: android/titanium/src/java/org/appcelerator/titanium/TiApplication.java (hunk 1 above)
Comment (jquick-axway) [reply android/titanium/src/java/org/appcelerator/titanium/TiApplication.java:621]: You're right. I'll change it. Thanks.

**Comment by ssekhri** [comment]
FR Passed
Mac OS: 10.15.4
SDK: 9.3.0.v20201016171621
Appc CLI: 8.1.1
JDK: 11.0.4
Node: 10.17.0
Studio: 6.0.0.202005141803
Device: Nexus4(v5.1.1) device, Pixel 3 XL(v11.0) emulator

**Pull request merged**